compiler/rustc_lint_defs/src/builtin.rs

```diff
@@ -2965,6 +2965,42 @@ declare_lint! {
     };
 }
 
+declare_lint! {
+    /// The `ptr_to_int_cast` lint checks casting pointers to integers other than `usize`. These
+    /// casts are often bugs.
+    ///
+    /// ### Example
+    ///
+    /// ```rust,compile_fail
+    /// #![deny(ptr_to_int_cast)]
+    ///
+    /// fn main() {
+    ///     let x = 100_000_000;
+    ///     let y = u16::max as u32; // a function pointer, cast to u32, causing an overflow
```

Comment:
```suggestion
    ///     let y = u16::max as u32; // the user meant the constant `u16::MAX`, rather than the
    ///                              // function `u16::max`, but this cast is technically permitted,
    ///                              // so will not produce an error
```